pkg/cloud/cloud.go

```diff
@@ -1,37 +1,296 @@
+	snapshotName := snapshotOptions.SnapshotName
+	volumeId := &snapshotOptions.SourceVolumeId
+	tags := snapshotOptions.Tags
+	if len(snapshotName) == 0 {
+		return nil, fmt.Errorf("snapshot name not provided")
+	}
+	if volumeId == nil {
+		return nil, fmt.Errorf("volume id not provided")
+	}
 
-	metadata, err := NewMetadataService(svc)
+	// The volume id associated with the persistent volume could refer to the file system id. In that case,
+	// we need to retrieve the id of the root volume associated with the file system
+	volumeId, err = c.getVolumeId(ctx, *volumeId)
 	if err != nil {
-		return nil, fmt.Errorf("could not get metadata from AWS: %v", err)
+		return nil, err
 	}
 
-	return &cloud{
-		metadata: metadata,
+	snapshotTags := []*fsx.Tag{
+		{
+			Key:   aws.String(SnapshotNameTagKey),
+			Value: aws.String(snapshotName),
+		},
+		{
+			Key:   aws.String(AwsFsxOpenZfsDriverTagKey),
+			Value: aws.String("true"),
+		},
+	}
+
+	if len(tags) != 0 {
+		snapshotTags = append(snapshotTags, util.ParseTags(tags)...)
+	}
+
+	input := &fsx.CreateSnapshotInput{
+		ClientRequestToken: aws.String(snapshotName),
+		Name:               aws.String(snapshotName),
+		Tags:               snapshotTags,
+		VolumeId:           volumeId,
+	}
+	klog.V(4).Infof("CreateSnapshotInput: ", input.GoString())
+	output, err := c.fsx.CreateSnapshotWithContext(ctx, input)
+
+	if err != nil {
+		return nil, fmt.Errorf("error creating snapshot of volume %s: %w", snapshotOptions.SourceVolumeId, err)
+	}
+	if output == nil {
+		return nil, fmt.Errorf("nil CreateSnapshotResponse")
+	}
+	klog.V(4).Infof("CreateSnapshotResponse: ", output.GoString())
+	return &Snapshot{
+		SnapshotID:     *output.Snapshot.SnapshotId,
+		SourceVolumeID: *volumeId,
+		CreationTime:   *output.Snapshot.CreationTime,
 	}, nil
 }
 
-func (c *cloud) GetMetadata() MetadataService {
-	return c.metadata
+func (c *cloud) DeleteSnapshot(ctx context.Context, snapshotId string) (err error) {
+	if len(snapshotId) == 0 {
+		return fmt.Errorf("snapshot id not provided")
+	}
+
+	input := &fsx.DeleteSnapshotInput{
+		ClientRequestToken: aws.String(snapshotId),
+		SnapshotId:         aws.String(snapshotId),
+	}
+	if _, err = c.fsx.DeleteSnapshotWithContext(ctx, input); err != nil {
+		if isSnapshotNotFound(err) {
+			return fmt.Errorf("DeleteSnapshot: Unable to find snapshot %s", snapshotId)
+		}
+		return fmt.Errorf("DeleteSnapshot: Failed to delete snapshot %s, received error %v", snapshotId, err)
+	}
+	return nil
+}
+
+func (c *cloud) WaitForSnapshotAvailable(ctx context.Context, snapshotId string) error {
+	if len(snapshotId) == 0 {
+		return fmt.Errorf("snapshot id not provided")
+	}
+
+	klog.V(2).Infof("WaitForSnapshotAvailable: Snapshot %s", snapshotId)
+	err := wait.Poll(SnapshotPollCheckInterval, PollCheckTimeout, func() (done bool, err error) {
+		klog.V(2).Infof("WaitForSnapshotAvailable: HELLO HUGHDAN %s", snapshotId)
```

Comment:
This seems like it was for debugging purposes, please remove

Reply:
Yikes - mb. I'll remove this